File: bench/python/ops/bench_rotate.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA Rotate operator benchmark - Python equivalent of BenchRotate.cpp"""

import math
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cupy as cp  # noqa: E402
import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    get_dtype,
    get_dtype_size,
    get_num_channels,
    get_interpolation_type,
    get_format_from_dtype,
    create_tensor,
    create_image_batch_varshape,
    create_stream_cache,
    run_benchmark,
)


def _rotate_in_bounds_fraction(angle_deg, x_shift, y_shift, w, h):
    """Fraction of dst pixels in [0,W]x[0,H] whose inverse rotation lands inside source.

    The rotate kernel skips both read and write outside this region (rotate.cu:61),
    so the byte model is scaled by this fraction to track HBM traffic.
    """
    th = math.radians(angle_deg)
    c, s = math.cos(th), math.sin(th)

    # Kernel mapping is src = R(+θ)·(dst - shift), so the in-bounds dst region is
    # the image of [0,W]x[0,H] under dst = R(-θ)·src + shift.
    def m(x, y):
        return (c * x + s * y + x_shift, -s * x + c * y + y_shift)

    poly = [m(0.0, 0.0), m(w, 0.0), m(w, h), m(0.0, h)]

    def clip(in_pts, axis, bound, keep_above):
        if not in_pts:
            return []
        out = []
        for i, cur in enumerate(in_pts):
            prev = in_pts[i - 1]
            ci = (cur[axis] >= bound) if keep_above else (cur[axis] <= bound)
            pi = (prev[axis] >= bound) if keep_above else (prev[axis] <= bound)
            if ci:
                if not pi:
                    t = (bound - prev[axis]) / (cur[axis] - prev[axis])
                    out.append(
                        (
                            prev[0] + t * (cur[0] - prev[0]),
                            prev[1] + t * (cur[1] - prev[1]),
                        )
                    )
                out.append(cur)
            elif pi:
                t = (bound - prev[axis]) / (cur[axis] - prev[axis])
                out.append(
                    (prev[0] + t * (cur[0] - prev[0]), prev[1] + t * (cur[1] - prev[1]))
                )
        return out

    poly = clip(poly, 0, 0.0, True)
    poly = clip(poly, 0, w, False)
    poly = clip(poly, 1, 0.0, True)
    poly = clip(poly, 1, h, False)
    if len(poly) < 3:
        return 0.0
    area = 0.0
    n = len(poly)
    for i in range(n):
        j = (i + 1) % n
        area += poly[i][0] * poly[j][1] - poly[j][0] * poly[i][1]
    return abs(area) * 0.5 / (w * h)


def rotate(state):
    """Rotate operator benchmark matching C++ BenchRotate.cpp"""

    shape = parse_shape(state.get_string("shape"))
    dtype_str = state.get_string("InOutDataType")
    _ = get_dtype(dtype_str)
    try:
        layout = state.get_string("layout")
    except Exception:
        layout = "NHWC"
    input_kind = get_input_kind(state.get_string("inputKind"))
    interp = get_interpolation_type(state.get_string("interpolation"))
    device_id = state.get_device()

    N, H, W = shape
    nc = get_num_channels(dtype_str)

    if layout not in ("NHWC", "NCHW", "NCHW_FAKE"):
        state.skip("Rotate benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts")
        return None

    is_planar = layout == "NCHW"
    is_fake_planar = layout == "NCHW_FAKE"
    if is_fake_planar and input_kind == "VarShape":
        state.skip("Fake-planar (NCHW_FAKE) rotate benchmark is tensor-only")
        return None

    # Rotation around image center: keeps the bulk of dst pixels inside source so
    # the kernel actually exercises read+interp+write rather than the out-of-bounds
    # early-exit guard at rotate.cu:61.
    angle_deg = 30.0
    th = math.radians(angle_deg)
    cx, cy = W * 0.5, H * 0.5
    x_shift = cx * (1.0 - math.cos(th)) - cy * math.sin(th)
    y_shift = cy * (1.0 - math.cos(th)) + cx * math.sin(th)
    shift = (x_shift, y_shift)

    in_bounds_frac = _rotate_in_bounds_fraction(angle_deg, x_shift, y_shift, W, H)
    bytes_full = N * H * W * get_dtype_size(dtype_str)
    # Native rotate only touches the in-bounds region; the two reformats in the fake-planar path move
    # the full tensor each way (reformat is full-coverage), so add 2*bytes_full on top of the rotate.
    bytes_io = int(
        (2 * bytes_full if is_fake_planar else 0) + in_bounds_frac * bytes_full
    )
    state.add_global_memory_reads(bytes_io)
    state.add_global_memory_writes(bytes_io)

    get_stream = create_stream_cache()

    if is_fake_planar:
        # Tensor-only "fake planar": planar->interleaved->rotate->interleaved->planar,
        # all timed, as the comparison baseline for the native planar (NCHW) path.
        src = create_tensor(
            (N, nc, H, W), dtype_str, device_id, layout="NCHW", fill_mode="lcg"
        )
        inter_src = create_tensor(
            (N, H, W, nc), dtype_str, device_id, layout="NHWC", fill_mode=0
        )
        inter_dst = create_tensor(
            (N, H, W, nc), dtype_str, device_id, layout="NHWC", fill_mode=0
        )
        dst = create_tensor(
            (N, nc, H, W), dtype_str, device_id, layout="NCHW", fill_mode=0
        )

        def run_fake(launch):
            stream = get_stream(launch)
            cvcuda.reformat_into(inter_src, src, stream=stream)
            cvcuda.rotate_into(
                inter_dst, inter_src, angle_deg, shift, interp, stream=stream
            )
            cvcuda.reformat_into(dst, inter_dst, stream=stream)

        return run_fake

    if input_kind == "Tensor":
        tshape = (N, nc, H, W) if is_planar else (N, H, W, nc)
        src = create_tensor(
            tshape, dtype_str, device_id, layout=layout, fill_mode="lcg"
        )
        dst = create_tensor(tshape, dtype_str, device_id, layout=layout, fill_mode=0)

        def run(launch):
            stream = get_stream(launch)
            cvcuda.rotate_into(dst, src, angle_deg, shift, interp, stream=stream)

        return run

    else:
        img_format = get_format_from_dtype(dtype_str, nc, planar=is_planar)
        src = create_image_batch_varshape(
            (N, H, W, nc),
            0,
            img_format,
            dtype=dtype_str,
            device=device_id,
            fill_mode="lcg",
        )
        dst = create_image_batch_varshape(
            (N, H, W, nc),
            0,
            img_format,
            dtype=dtype_str,
            device=device_id,
            fill_mode=0,
        )

        with cp.cuda.Device(device_id):
            angle_data = cp.full((N,), angle_deg, dtype=cp.float64)
            shift_data = cp.zeros((N, 2), dtype=cp.float64)
            shift_data[:, 0] = shift[0]
            shift_data[:, 1] = shift[1]
        angle_tensor = cvcuda.as_tensor(angle_data, "N")
        shift_tensor = cvcuda.as_tensor(shift_data, "NW")

        def run(launch):
            stream = get_stream(launch)
            cvcuda.rotate_into(
                dst, src, angle_tensor, shift_tensor, interp, stream=stream
            )

        return run


if __name__ == "__main__":
    run_benchmark("rotate", rotate)

File: bench/cpp/ops/BenchRotate.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchRotateConfig.hpp"

#include <cvcuda/OpReformat.hpp>
#include <cvcuda/OpRotate.hpp>

#include <nvbench/nvbench.cuh>

#include <cmath>
#include <vector>

namespace {

// Sutherland-Hodgman half-plane clip.  axis=0 clips on x, axis=1 on y.
// keepAbove=true keeps points whose coord >= bound; false keeps coord <= bound.
inline std::vector<double2> clipHalfPlane(const std::vector<double2> &in, int axis, double bound, bool keepAbove)
{
    std::vector<double2> out;
    if (in.empty())
        return out;
    auto coord = [axis](double2 p)
    {
        return axis == 0 ? p.x : p.y;
    };
    auto inside = [&coord, &keepAbove, &bound](double2 p)
    {
        return keepAbove ? coord(p) >= bound : coord(p) <= bound;
    };
    auto inter = [&coord, &bound](double2 a, double2 b) -> double2
    {
        double t = (bound - coord(a)) / (coord(b) - coord(a));
        return {a.x + t * (b.x - a.x), a.y + t * (b.y - a.y)};
    };
    for (size_t i = 0; i < in.size(); ++i)
    {
        double2 cur  = in[i];
        double2 prev = in[(i + in.size() - 1) % in.size()];
        bool    ci   = inside(cur);
        bool    pi   = inside(prev);
        if (ci)
        {
            if (!pi)
                out.push_back(inter(prev, cur));
            out.push_back(cur);
        }
        else if (pi)
        {
            out.push_back(inter(prev, cur));
        }
    }
    return out;
}

// Fraction of dst pixels in [0,W]x[0,H] whose inverse rotation maps inside source.
// The rotate kernel skips both read and write outside this region, so the byte model
// must be scaled by this fraction to track HBM traffic instead of nominal coverage.
inline double rotateInBoundsFraction(double angleDeg, double xShift, double yShift, double W, double H)
{
    const double th = angleDeg * M_PI / 180.0;
    const double c  = std::cos(th);
    const double s  = std::sin(th);
    // Kernel mapping is src = R(+θ) * (dst - shift), so the in-bounds dst region is
    // the image of [0,W]x[0,H] under  dst = R(-θ) * src + shift.
    auto         map = [&c, &s, &xShift, &yShift](double x, double y)
    {
        return double2{c * x + s * y + xShift, -s * x + c * y + yShift};
    };
    std::vector<double2> p = {map(0.0, 0.0), map(W, 0.0), map(W, H), map(0.0, H)};
    p                      = clipHalfPlane(p, 0, 0.0, true);
    p                      = clipHalfPlane(p, 0, W, false);
    p                      = clipHalfPlane(p, 1, 0.0, true);
    p                      = clipHalfPlane(p, 1, H, false);
    if (p.size() < 3)
        return 0.0;
    double area = 0.0;
    for (size_t i = 0; i < p.size(); ++i)
    {
        size_t j = (i + 1) % p.size();
        area += p[i].x * p[j].y - p[j].x * p[i].y;
    }
    return std::abs(area) * 0.5 / (W * H);
}

} // namespace

template<typename T>
inline void rotate(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape     = benchutils::GetShape<3, int3>(state.get_string("shape"));
    auto                        layout    = benchutils::GetStringParam(state, "layout", "NHWC");
    const benchutils::InputKind inputKind = benchutils::GetInputKind(state.get_string("inputKind"));

    NVCVInterpolationType interpType = benchutils::GetInterpolationType(state.get_string("interpolation"));

    const bool isPlanar     = layout == "NCHW";
    const bool isFakePlanar = layout == "NCHW_FAKE";
    if (layout != "NHWC" && layout != "NCHW" && layout != "NCHW_FAKE")
    {
        state.skip("Rotate benchmark supports only NHWC, NCHW, and NCHW_FAKE layouts");
        return;
    }
    // NCHW_FAKE ("fake planar") is a tensor-only comparison path: planar data is reformatted to
    // interleaved, rotated with the interleaved kernel, and reformatted back to planar — all timed
    // together — so the native planar path (NCHW) can be shown to be faster than this naive
    // convert→rotate→convert pipeline.
    if (isFakePlanar && inputKind == benchutils::InputKind::VarShape)
    {
        state.skip("Fake-planar (NCHW_FAKE) rotate benchmark is tensor-only");
        return;
    }

    // Rotation around image center: keeps the bulk of dst pixels inside source so
    // the kernel actually exercises the read+interp+write path rather than the
    // out-of-bounds early-exit guard at rotate.cu:61.
    const double angleDeg = 30.0;
    const double th       = angleDeg * M_PI / 180.0;
    const double cx       = static_cast<double>(shape.z) * 0.5;
    const double cy       = static_cast<double>(shape.y) * 0.5;
    const double xShift   = cx * (1.0 - std::cos(th)) - cy * std::sin(th);
    const double yShift   = cy * (1.0 - std::cos(th)) + cx * std::sin(th);
    double2      shift{xShift, yShift};

    const double inBoundsFrac
        = rotateInBoundsFraction(angleDeg, xShift, yShift, static_cast<double>(shape.z), static_cast<double>(shape.y));
    const double fullBytes = static_cast<double>(shape.x) * shape.y * shape.z * sizeof(T);
    // Native rotate only touches the in-bounds region; the two reformats in the fake-planar path move
    // the full tensor each way (reformat is full-coverage), so add 2*fullBytes on top of the rotate.
    const auto   bytesIO = static_cast<size_t>((isFakePlanar ? 2.0 * fullBytes : 0.0) + inBoundsFrac * fullBytes);
    state.add_global_memory_reads(bytesIO);
    state.add_global_memory_writes(bytesIO);

    cvcuda::Rotate op(shape.x);

    // clang-format off

    if (isFakePlanar) // tensor-only: planar→interleaved→rotate→interleaved→planar
    {
        using BT = typename nvcv::cuda::BaseType<T>;
        int  ch  = nvcv::cuda::NumElements<T>;

        nvcv::Tensor src     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interSrc({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor interDst({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst     ({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::LcgValues<BT>());

        cvcuda::Reformat reformatOp;

        benchutils::warmup_and_exec(state, BENCH_ROTATE_WARMUP_ITERATIONS,
            [&op, &reformatOp, &src, &interSrc, &interDst, &dst, &angleDeg, &shift, &interpType](cudaStream_t s) {
                reformatOp(s, src, interSrc);                       // NCHW → NHWC
                op(s, interSrc, interDst, angleDeg, shift, interpType); // interleaved rotate
                reformatOp(s, interDst, dst);                       // NHWC → NCHW
            });
    }
    else if (inputKind == benchutils::InputKind::Tensor)
    {
        using BT = typename nvcv::cuda::BaseType<T>;
        int  ch  = nvcv::cuda::NumElements<T>;

        nvcv::Tensor src = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());
        nvcv::Tensor dst = isPlanar
                             ? nvcv::Tensor({{shape.x, ch, shape.y, shape.z}, "NCHW"}, benchutils::GetDataType<BT>())
                             : nvcv::Tensor({{shape.x, shape.y, shape.z, ch}, "NHWC"}, benchutils::GetDataType<BT>());

        benchutils::FillTensor<BT>(src, benchutils::LcgValues<BT>());

        benchutils::warmup_and_exec(state, BENCH_ROTATE_WARMUP_ITERATIONS,
            [&op, &src, &dst, &angleDeg, &shift, &interpType](cudaStream_t s) {
                op(s, src, dst, angleDeg, shift, interpType);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        nvcv::ImageBatchVarShape src(shape.x);
        nvcv::ImageBatchVarShape dst(shape.x);

        if (isPlanar)
        {
            benchutils::FillPlanarImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0});
            benchutils::FillPlanarImageBatch<T>(dst, long2{shape.z, shape.y}, long2{0, 0});
        }
        else
        {
            benchutils::FillImageBatch<T>(src, long2{shape.z, shape.y}, long2{0, 0},
                                          benchutils::LcgValues<T>());
            // Use FillImageBatchLike to ensure dst has same per-sample shapes as src
            benchutils::FillImageBatchLike<T>(dst, src, [](const long4_16a &) { return T{0}; });
        }

        nvcv::Tensor angleDegTensor({{shape.x}, "N"}, nvcv::TYPE_F64);
        nvcv::Tensor shiftTensor({{shape.x, 2}, "NW"}, nvcv::TYPE_F64);

        benchutils::FillTensor<double>(angleDegTensor, [&angleDeg](const long4_16a &){ return angleDeg; });
        benchutils::FillTensor<double>(
            shiftTensor, [&shift](const long4_16a &c) { return nvcv::cuda::GetElement(shift, static_cast<int>(c.y)); });

        benchutils::warmup_and_exec(state, BENCH_ROTATE_WARMUP_ITERATIONS,
            [&op, &src, &dst, &angleDegTensor, &shiftTensor, &interpType](cudaStream_t s) {
                op(s, src, dst, angleDegTensor, shiftTensor, interpType);
            });
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(rotate, NVBENCH_TYPE_AXES(BENCH_ROTATE_TYPES))
BENCH_ROTATE_AXES;
